### PR TITLE
Fix for correct exec command arguments

### PR DESCRIPTION
**Pull request by macedot**
Last code didn't pass custom ini files for acServer at the command line.

This commit also moves ini creation for /config/<instance id> instead of cfg (const).

json can be ignored. (sry :\)

### CODE REVIEW

**Comment by DeKugelschieber** [comment]
Have you actually tested your changes?

**Comment by macedot** [comment]
Yes i had.
http://i.imgur.com/JHzek0b.png

**Comment by DeKugelschieber** [comment]
Yeah I have seen that, but your pull request fixes some issues with the current version, so I'll merge it tomorrow.

**Comment by macedot** [comment]
I've adapted venyii pull to #24. 

I've also changed cfg config and ini path generation to exported methods inside instance package.

File: tools/tracks.json [hunk 1]
```diff
@@ -0,0 +1,272 @@
+[
```
Comment (DeKugelschieber) [inline tools/tracks.json]: Remove this file from Git and keep tools/ clean. And remove tools.exe as well.

File: src/rest/api.go [hunk 2]
```diff
@@ -346,6 +353,67 @@ func GetConfiguration(w http.ResponseWriter, r *http.Request) {
...
+	_, err = fw.Write(dat)
+	if err != nil {
+		log.Printf("Error writing %s config to zip: %v", filename, err)
+		return false
+	}
+
+	return true
+}
+
+func downloadConfiguration(w http.ResponseWriter, r *http.Request) {
```
Comment (DeKugelschieber) [inline src/rest/api.go]: Rename to downloadConfigurationHandler and outsource logic (except for the part that does the parsing and rewriting header) into a new file inside the configuration package. AddFileToZip should be outsourced as well. The rest package should stay as small as possible.

File: src/rest/api.go [hunk 3]
```diff
@@ -346,6 +353,67 @@ func GetConfiguration(w http.ResponseWriter, r *http.Request) {
...
+	id, err := strconv.Atoi(r.URL.Query().Get("id"))
+
+	if err != nil {
+		resp.Error(w, 100, err.Error(), nil)
+		return
+	}
+
+	config, err := config.GetConfiguration(int64(id))
```
Comment (DeKugelschieber) [inline src/rest/api.go]: After outsourcing, this should look like: `zipFile, err := config.GetAndZipConfiguration(int64(id))`

File: src/instance/mgn.go [hunk 4]
```diff
@@ -53,13 +53,18 @@ func StartInstance(name string, configuration int64) error {
...
+
+	if err != nil {
 		return util.OpError{3, "Error writing configuration"}
 	}
 
+	// force server_cfg and entry_list ini paths
 	// start
-	cmd := exec.Command(filepath.Join(s.Folder, s.Executable), strings.Split(s.Args, " ")...)
-	now := strings.Replace(time.Now().String(), " ", "_", -1)
+	// FIXME: s.Args has been  discarted. No real use so far?
```
Comment (DeKugelschieber) [inline src/instance/mgn.go]: s.Args is used to pass start parameters from the UI (like `linux` shown in the demo video: https://www.youtube.com/watch?v=z6osf2BMKqE). Please add it as the last parameter to exec.Command(... (the split is required!).

**Comment by DeKugelschieber** [comment]
Maybe you want to add an error message to the UI if an error occures while creating the zip file on download.

**Comment by venyii** [comment]
Umm.... why was #25 cherry-picked (and changed) in this PR?

**Comment by DeKugelschieber** [comment]
I don't know, maybe @macedot should just remove that one commit from this branch, if still possible.

**Comment by macedot** [comment]
Because @venyii changed write ini into a create ini buffer. Even though the same logic was applied to both download and save process, it would not be consistent: downloading ini files would result in "downloading reconstructed ini", not current ini in use.

Say you've started a server with config (A0)
Then you change something at config A so now it differs (A1)
Using @venyii commit, download process would reconstruct A1, but current instance is running with A0.

From my perspective, user would like to **Download** currently in-use ini, not **Export** most up-to-date instance config.

File: src/instance/mgn.go (hunk 4 above)
Comment (macedot) [inline src/instance/mgn.go]: I keep my question: does it reflect an real use case? Because acServer doesn't need **linux** parameter.

**Comment by venyii** [comment]
My plan was to download the latest saved version of a configuration which is 'rebuilt' fresh from the DB on every download.

**Comment by DeKugelschieber** [comment]
We could add a download button on the configuration and on the instances page. The one on the instances page downloads the configuration in use and the one on the configuration page downloads the lastest configuration.

File: .dockerignore [hunk 5]
```diff
@@ -14,4 +14,5 @@ REAMDE.md
 aclogo.png
 build_release
 run
-
+tools/*.json
```
Comment (DeKugelschieber) [inline .dockerignore]: You may want to add this to the .gitignore as well.

File: src/instance/config.go [hunk 6]
```diff
@@ -11,31 +11,47 @@ import (
 )
 
 const (
-	cfg_folder     = "cfg"
+	// cfg_folder     = "cfg"
```
Comment (DeKugelschieber) [inline src/instance/config.go]: Remove it if not used.

File: src/instance/mgn.go (hunk 4 above)
Comment (DeKugelschieber) [inline src/instance/mgn.go]: I don't know, maybe there are some parameters that someone wants to add. It makes it a little more flexible.

File: src/rest/api.go (hunk 2 above)
Comment (DeKugelschieber) [inline src/rest/api.go]: @venyii

File: src/rest/api.go (hunk 3 above)
Comment (DeKugelschieber) [inline src/rest/api.go]: @venyii

File: src/rest/api.go (hunk 2 above)
Comment (venyii) [inline src/rest/api.go]: Do you want me to do these comments in my own PR? I can't do anything in this PR and since it's not mine and the commit was already changed. It would only mean more work for @macedot.

Comment (DeKugelschieber) [inline src/rest/api.go]: The PRs should have stayed separated in the first place. So do it on your own PR and macedot needs to merge it.

File: src/instance/mgn.go (hunk 4 above)
Comment (macedot) [inline src/instance/mgn.go]: Agree. Go is totally new for me (those are my first attempts on it). I've been fooled by exec.Command enough to discard joining ini parameters with s.Args. :o)
Will try to add it after refactor zip thing.

File: src/rest/api.go (hunk 2 above)
Comment (macedot) [inline src/rest/api.go]: Ops :\

**Comment by macedot** [comment]
I've finished coding merge from @venyii commit. I've reverted back into its code and extended the interface as requested.
I've also created the button into instance page to download current instance files using @venyii code.
I'm finishing tests on it. should be done soon.

**Comment by macedot** [comment]
Finally, it is done :)

i hereby declare to whom it may concern that i'll stop making changes in master without branching first.

:o)

File: src/instance/config.go [hunk 7]
```diff
@@ -11,31 +11,63 @@ import (
...
+}
+
+func GetEntryListPath(config *model.Configuration) string {
+	iniFile := filepath.Join(GetConfigPath(config), entry_list_ini)
+	return iniFile
+}
+
+func writeIniFile(config *model.Configuration, ini, filename string) error {
+	if err := ioutil.WriteFile(filename, []byte(ini), 0775); err != nil {
+		log.WithFields(log.Fields{"err": err}).Errorf("Error writing %s", filename)
```
Comment (DeKugelschieber) [inline src/instance/config.go]: Remove the `... %s", filename` and user Error(). Additional information should be added as fields: `log.WithFields(log.Fields{"err": err, "filename": filename}).Error(...`.

File: src/instance/config.go [hunk 8]
```diff
@@ -11,31 +11,63 @@ import (
...
 	server_ini     = "server_cfg.ini"
 	entry_list_ini = "entry_list.ini"
 	sep            = "\n"
 )
 
-func writeConfig(config *model.Configuration) error {
-	iniPath := filepath.Join(cfg_folder, int64ToStr(config.Id))
-	if err := os.MkdirAll(iniPath, 0755); err != nil {
-		log.WithFields(log.Fields{"err": err}).Error("Error creating cfg folder")
+func GetServerCfgFileName() string {
```
Comment (DeKugelschieber) [inline src/instance/config.go]: Remove the function and just use the constant.

File: src/instance/config.go [hunk 9]
```diff
@@ -11,31 +11,63 @@ import (
...
+	return server_ini
+}
+
+func GetEntryListFileName() string {
```
Comment (DeKugelschieber) [inline src/instance/config.go]: Remove the function and just use the constant.

File: src/rest/api.go [hunk 10]
```diff
@@ -3,15 +3,16 @@ package rest
...
-	log "github.com/sirupsen/logrus"
 	"instance"
 	"model"
 	"net/http"
 	"settings"
 	"strconv"
 	"user"
+
+	resp "github.com/DeKugelschieber/go-resp"
+	session "github.com/DeKugelschieber/go-session"
```
Comment (DeKugelschieber) [inline src/rest/api.go]: `resp` and `session` is not required, since those are the names of the packages already.

File: src/rest/api.go [hunk 11]
```diff
@@ -346,6 +354,38 @@ func GetConfiguration(w http.ResponseWriter, r *http.Request) {
...
 
+func downloadConfigurationHandler(w http.ResponseWriter, r *http.Request, dlType string) error {
+	id, err := strconv.Atoi(r.URL.Query().Get("id"))
+	if err != nil {
+		resp.Error(w, 100, err.Error(), nil)
+		return err
+	}
+
+	config, err := config.GetConfiguration(int64(id))
+	if err != nil {
```
Comment (DeKugelschieber) [inline src/rest/api.go]: Use `iserror(err)` and just return nothing.

File: src/rest/api.go [hunk 12]
```diff
@@ -346,6 +354,38 @@ func GetConfiguration(w http.ResponseWriter, r *http.Request) {
...
+
+	w.Header().Set("Content-Disposition", "attachment; filename=\""+config.Name+".zip\"")
+	w.Header().Set("Content-Type", "application/zip")
+
+	if dlType == "1" {
+		err = instance.ZipConfiguration(config, w)
+	} else if dlType == "2" {
+		err = instance.ZipInstanceFiles(config, w)
+	} else {
+		return nil
```
Comment (DeKugelschieber) [inline src/rest/api.go]: Return an error code to the client and nothing here.

File: src/rest/api.go [hunk 13]
```diff
@@ -346,6 +354,38 @@ func GetConfiguration(w http.ResponseWriter, r *http.Request) {
 	w.Write(resp)
 }
 
+func downloadConfigurationHandler(w http.ResponseWriter, r *http.Request, dlType string) error {
```
Comment (DeKugelschieber) [inline src/rest/api.go]: Don't return error in HTTP handlers.

File: src/rest/api.go [hunk 14]
```diff
@@ -346,6 +354,38 @@ func GetConfiguration(w http.ResponseWriter, r *http.Request) {
...
+
+	if dlType == "1" {
+		err = instance.ZipConfiguration(config, w)
+	} else if dlType == "2" {
+		err = instance.ZipInstanceFiles(config, w)
+	} else {
+		return nil
+	}
+
+	if err != nil {
```
Comment (DeKugelschieber) [inline src/rest/api.go]: Use `iserror(err)` and just return nothing.

File: src/rest/api.go [hunk 15]
```diff
@@ -346,6 +354,38 @@ func GetConfiguration(w http.ResponseWriter, r *http.Request) {
...
+		resp.Error(w, 100, err.Error(), nil)
+		return err
+	}
+
+	return nil
```
Comment (DeKugelschieber) [inline src/rest/api.go]: Use `success(w)` instead.

File: src/instance/config.go (hunk 9 above)
Comment (macedot) [inline src/instance/config.go]: Thus, constants will be renamed. Continue?
(y/n)

Comment (DeKugelschieber) [inline src/instance/config.go]: Yeah go ahead :)

**Comment by macedot** [comment]
Job's done. :)

**Comment by venyii** [comment]
Thanks for the work @macedot and continuing the download PR 👍
